apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.cs

```diff
@@ -255,13 +255,117 @@ public static SpannerCommandTextBuilder FromCommandText(string commandText)
+                                "https://cloud.google.com/spanner/docs/dml-versus-mutations.");
+                        }
                         targetTable = ValidateTableName(commandSections[1], nameof(commandText));
                         // This just normalizes the command text.
                         commandText = $"{commandName} {targetTable}";
                     }
                     break;
             }
-            return new SpannerCommandTextBuilder(commandText, commandType, targetTable, extraStatements: null);
+            return new SpannerCommandTextBuilder(commandText.Trim(), commandType, targetTable, extraStatements: null);
+        }
+
+        private static string RemoveLeadingCommentsAndHints(string commandText, out bool removedCommentOrHint)
+        {
+            try
+            {
+                return commandText.Substring(IndexOfCommandStart(commandText, out removedCommentOrHint));
+            }
+            catch(Exception ex) when (ex is ArgumentOutOfRangeException || ex is IndexOutOfRangeException)
+            {
+                throw new ArgumentException($"'{commandText}' is not a recognized Spanner command.", nameof(commandText), ex);
+            }
+        }
+
+        private static int IndexOfCommandStart(string command, out bool removedCommentOrHint)
+        {
+            int currentIndex = 0;
+            int loops = 0;
+            do
+            {
+                loops++;
+                SkipWhiteSpaces();
+            }
+            while (SkipLineComment() || SkipBlockComment() || SkipHint());
+
+            removedCommentOrHint = loops > 1;
+            return currentIndex;
+
+            void SkipWhiteSpaces()
+            {
+                while (char.IsWhiteSpace(command[currentIndex]))
+                {
+                    currentIndex++;
+                }
+            }
+
+            bool SkipLineComment()
+            {
+                if (command.IndexOf('#', currentIndex, 1) >= 0 ||
+                    command.IndexOf("--", currentIndex, 2) >= 0)
+                {
+                    // TODO: If this can be escaped, let's check that it's not
+                    // or continue searching.
+                    currentIndex = command.IndexOf('\n', currentIndex);
+                    if (currentIndex >= 0) // If we didn't find it, we don't fail here.
+                    {
+                        currentIndex++;
+                    }
+                    return true;
+                }
+                return false;
+            }
+
+            bool SkipBlockComment()
+            {
+                if (command.IndexOf("/*", currentIndex, 2) >= 0 )
+                {
+                    // TODO: If this can be escaped, let's check that it's not
+                    // or continue searching.
+                    currentIndex = command.IndexOf("*/", currentIndex + 2);
+                    if (currentIndex >= 0) // If we didn't find it, we don't fail here.
```

Comment:
Again, presumably this is "Continue the search from the end of the comment"?

If we *don't* find the end of the comment, presumably now currentIndex is -1 and we'll throw an exception accidentally?

Reply:
Ah - I'd missed that we're already catching those exceptions. I think I'd *personally* prefer to remove that catch block and instead throw dedicated exceptions... I'm slightly worried that if we've got a bug in this code, we'll end up hiding it. We should be able to work out every case where we *would* end up throwing an exception, and doing so explicitly instead. (I don't mind if it throws an ArgumentException with "unrecognized command" as the message - it's more about doing it explicitly.)